app/apps/server/storage/AppGridFSSourceStorage.ts

```diff
@@ -0,0 +1,75 @@
+import { MongoInternals } from 'meteor/mongo';
+import { GridFSBucket, GridFSBucketWriteStream, ObjectId } from 'mongodb';
+import { AppSourceStorage, IAppStorageItem } from '@rocket.chat/apps-engine/server/storage';
+
+import { streamToBuffer } from '../../../file-upload/server/lib/streamToBuffer';
+
+export class AppGridFSSourceStorage extends AppSourceStorage {
+	private pathPrefix = 'GridFS:/';
+
+	private bucket: GridFSBucket;
+
+	constructor() {
+		super();
+
+		const { GridFSBucket } = MongoInternals.NpmModules.mongodb.module;
+		const { db } = MongoInternals.defaultRemoteCollectionDriver().mongo;
+
+		this.bucket = new GridFSBucket(db, {
+			bucketName: 'rocketchat_apps_packages',
+			chunkSizeBytes: 1024 * 255,
+		});
+	}
+
+	public async store(item: IAppStorageItem, zip: Buffer): Promise<string> {
+		return new Promise((resolve, reject) => {
+			const filename = this.itemToFilename(item);
+			const writeStream: GridFSBucketWriteStream = this.bucket.openUploadStream(filename)
+				.on('finish', () => resolve(this.idToPath(writeStream.id)))
+				.on('error', (error) => reject(error));
+
+			writeStream.write(zip);
+			writeStream.end();
+		});
+	}
+
+	public async fetch(item: IAppStorageItem): Promise<Buffer> {
+		return streamToBuffer(this.bucket.openDownloadStream(this.itemToObjectId(item)));
+	}
+
+	public async update(item: IAppStorageItem, zip: Buffer): Promise<string> {
```

Comment:
As it is here, the `update` function will create a new file in GridFS with the new version of the app, while the previous version will be left "orphaned". This is not how it works in the FS counterpart.

What do you think should be the standardized behavior? @thassiov

Reply:
I think storing the new one, changing the reference in the db and then deleting the old one afterwards would be the way to go because we could for instance revert to the old one if there's some scenario we couldn't solve, like if the directory/file gets its permissions changed, or the server shuts down in the middle of it and we end up with some messy data